source/scripts/data/monsters.js

```diff
@@ -31,4 +31,206 @@ export default {
+                    delete choices[choices.indexOf({x: +1})]
+            }
+            for(var monster of this.game.monsters) {
+                if (monster.position.y == this.position.y - 1 && monster.position.x == this.position.x)
+                    delete choices[choices.indexOf({y: -1})]
+            }
+            for(var monster of this.game.monsters) {
+                if (monster.position.y == this.position.y + 1 && monster.position.x == this.position.x)
+                    delete choices[choices.indexOf({y: +1})]
+            }
+
+            if (this.position.x == 0)
+                delete choices[choices.indexOf({x: -1})]
+            if (this.position.x == DATA.FRAME.WIDTH)
+                delete choices[choices.indexOf({x: +1})]
+            if (this.position.y == 0)
+                delete choices[choices.indexOf({y: -1})]
+            if (this.position.y == DATA.FRAME.HEIGHT)
+                delete choices[choices.indexOf({y: +1})]
+
+            return choices[Math.floor((Math.random() * choices.length))]
+        }
+    },
+    BLUE_BAT: {
+        sprite: DATA.SPRITES.MONSTERS.BAT,
+        color: DATA.COLORS.BLUE,
+        health: 1,
+        strength: 1,
+        movement: function () {
+            if (this.position.x < 0) return {x: +1}
+            if (this.position.x >= DATA.FRAME.WIDTH) return {x: -1}
+            if (this.position.y < 0) return {y: +1}
+            if (this.position.y >= DATA.FRAME.HEIGHT) return {y: -1}
+
+            var choices = [
+                {x: -1},
+                {x: +1},
+                {y: -1},
+                {y: +1}
+            ]
+
+            for(var monster of this.game.monsters) {
+                if (monster.position.x == this.position.x - 1 && monster.position.y == this.position.y)
+                    delete choices[choices.indexOf({x: -1})]
+            }
+            for(var monster of this.game.monsters) {
+                if (monster.position.x == this.position.x + 1 && monster.position.y == this.position.y)
+                    delete choices[choices.indexOf({x: +1})]
+            }
+            for(var monster of this.game.monsters) {
+                if (monster.position.y == this.position.y - 1 && monster.position.x == this.position.x)
+                    delete choices[choices.indexOf({y: -1})]
+            }
+            for(var monster of this.game.monsters) {
+                if (monster.position.y == this.position.y + 1 && monster.position.x == this.position.x)
+                    delete choices[choices.indexOf({y: +1})]
+            }
+
+            if (this.position.x == 0)
+                delete choices[choices.indexOf({x: -1})]
+            if (this.position.x == DATA.FRAME.WIDTH)
+                delete choices[choices.indexOf({x: +1})]
+            if (this.position.y == 0)
+                delete choices[choices.indexOf({y: -1})]
+            if (this.position.y == DATA.FRAME.HEIGHT)
+                delete choices[choices.indexOf({y: +1})]
```

Comment:
The tricky part for this is how it is handled differently by different bat types. The only way I can think of that would work for all bat types is as follows (but it seems like it could be slow):
- have a function that gets an array of all empty tiles (free of walls, monsters, player, etc)
- in the movement for a particular monster, have an array of available movement vectors
- during the movement for a particular monster, find intersections of those two arrays and randomly choose from that.

The downside with that is having to generate a free/not-free map of the whole field for every movement. I couldn't find a way to directly access a tile to see if a monster stood on it (eg. something like `this.game.monsters["5x2"]` won't work due to the way monsters are keyed). That means that we'd have to build up an array of free locations by checking for terrain, then remove entries as we loop over all monsters.